### PR TITLE
Users can now purchase stories through Stripe

### PR DESCRIPTION
**Pull request by behica**
Finally. Got Stripe working, but took me a looooooooooooooooooooooooooooooooong time to figure out how to make it so that the `@amount` is variable. Ended up pretty simple by adding a hidden field on the form that brings in `story_id` and having the `ChargesController` find the story before doing anything to set the price.

This isn't what I had envisioned when I first thought of the app, since right now people are just purchasing stories from the site owner aka me, but it's a good first step I think. I'll have to figure out rights management and peer-to-peer payments for stories if I want it to be a real self-publishing platform / marketplace. Perhaps this can be handled through the donation model instead, where users donate to other users directly instead of having story tiers and paid stories.

### CODE REVIEW

**Comment by behica** [comment]
Also added purchase model. Story is now added to current_user purchases so that user cannot buy story over and over (view changes based on if story is in purchases). Also added functionality so that paid stories get automatically added to user's collection.

File: app/controllers/charges_controller.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+require 'stripe'
+class ChargesController < ApplicationController
+  before_action :authenticate_user!
+  before_action :find_story
+  
+  def new
+
```
Comment (bobbilee19) [inline app/controllers/charges_controller.rb]: remove space

File: app/controllers/charges_controller.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    )
+    
+    charge = Stripe::Charge.create(
+      customer: customer.id,
+      amount: 	@amount,
+      description: "Purchase of #{@story.title} for #{current_user.email}",
+      currency: 'usd'
+    )
+    
+    #current_user.update(customer_id: customer.id)
```
Comment (bobbilee19) [inline app/controllers/charges_controller.rb]: remove commented out stuff if not using for now, u will  have in git history if u ned it back

File: app/views/charges/_charge.html.erb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+      </div>
+    <% end %>
+    <label class="amount">
+      <span>Amount: <%= number_to_currency(@price) %></span>
+    </label>
+  </article>
+
+  <script class='stripe-button' src="https://checkout.stripe.com/checkout.js" 
+  data-key="<%= Rails.configuration.stripe[:publishable_key] %>" 
+  data-amount="<%= @story.price.to_i * 100 %>"
```
Comment (bobbilee19) [inline app/views/charges/_charge.html.erb]: do this stuff in controller

File: app/views/stories/new.html.erb [hunk 4]
```diff
@@ -24,26 +24,26 @@
...
+          </div>
+          <div class="form-group col-md-2">
+              <%= f.label :price %>
+              <%= f.number_field :price, class: 'form-control', placeholder: "Free" %>
+          </div>
         </div>
+        <%= f.submit "Save", class: 'btn btn-success' %>
+        <!--<div class="form-group">-->
+        <!--  <%= f.label :published, class: 'checkbox-inline' do %>-->
+        <!--    <%= f.check_box :published %><p class='bg-danger'> Publish - Story is submitted and cannot be edited! </p>-->
```
Comment (bobbilee19) [inline app/views/stories/new.html.erb]: remove some of the commented out stuff too when not using u will find in git history when needed

**Comment by bobbilee19** [comment]
Overall looks great! just a couple comments

File: app/views/charges/_charge.html.erb (hunk 3 above)
Comment (behica) [inline app/views/charges/_charge.html.erb]: Do you mean put the rails variables in the controller or the entire <script> itself?

File: app/controllers/charges_controller.rb (hunk 2 above)
Comment (behica) [inline app/controllers/charges_controller.rb]: Will remove for next commit.

File: app/views/stories/show.html.erb [hunk 5]
```diff
@@ -22,7 +22,11 @@
         <p><%= render partial: 'ratings/rate', locals: { story: @story } %></p>
       <% end %>
       <% if @story.tier == 'paid' %>
-        <%= render 'charges/charge', story_id: @story.id %>
+        <% if current_user && current_user.purchase?(@story) %>
```
Comment (behica) [inline app/views/stories/show.html.erb]: I need to re-organize these permissions, left it as is for now to show it all working. Ideally a user won't be able to add paid stories to their collection or rate them until they purchase it. I'll do a re-draw of the show page eventually.

**Pull request merged**